053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,452 @@
+   would predict that just reading from the notification bus could be faster
+   (especially considering the current proceduce is: i) receive empty event from
+   notification bus, and ii) perform a query) - this is just a hypothesis, though.
+   That said, other backends (e.g. an S3 backend) could be relatively slow to `Get`.
+
+Note that `EventStore` implementations don't need to think about notifications.
+This all happens at the DB layer:
+
+```go
+func (b *build) SaveEvent(ctx context.Context, event atc.Event) error {
+    // use the `EventStore` to save
+    key, err := b.eventStore.Put(ctx, b, event)
+    if err != nil {
+        return err
+    }
+    // ...but notify the main Postgres DB's notification bus.
+    payload := constructPayload(event, key)
+    return b.conn.Bus().Notify(buildEventsNotificationChannel(b.ID()), payload)
+}
+```
+
+```go
+func (b *build) Events(ctx context.Context, from uint) (EventSource, error) {
+    notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsNotificationChannel(b.ID()), func() (bool, error) {
+        return true, nil
+    })
+    if err != nil {
+        return nil, err
+    }
+
+    return newBuildEventSource(
+        ctx,
+        b,
+        b.conn,
+        b.eventStore, // fetch existing events from the EventStore using EventStore.Get(...)
+        notifier,     // but subscribe to the Postgres notification bus to receive new events
+        from,
+    ), nil
+}
+```
+
+
+## <a name="migrations">Migrations</a>
+
+Currently, we have migrations to set up the `build_events` table, as well as
+`TRIGGER`s for creating/deleting `pipeline_build_events_x` and
+`team_build_events_x` tables when pipelines and teams are created/deleted.
+
+I propose that we comment out these up migrations (so future deployments don't
+get these tables/triggers), and leave any database initialization (or whatever
+appropriate backend setup) to `BuildEvent.Setup()`. This means `Setup` will
+have to be "smart" (i.e. have to handle the cases where the initialization has
+already been applied). It also makes it difficult to make changes to the
+schema, unlike with migrations - each `EventStore` (that stores data in a
+structured way - this may just end up being Postgres) will need to basically
+implement its own form of migrations if we need to change the schema. My guess
+is that we won't be changing the schema too often, though (given that we
+haven't changed `build_events` since the initial migration).
+
+
+## Configuration
+
+One thing I haven't really explored is how operators would configure the
+backend store. I think it could be similar to how Credential managers are
+configured - there are groups of independent configuration options for each
+possible `EventStore`. If certain flags are configured for a given backend, use
+that backend.
+
+
+# Open Questions
+
+* Does `Key` need to be `interface{...}`, or can it be more specific like `uint`?
```

Comment:
In my [experimental PR](https://github.com/concourse/concourse/pull/5651) that introduced an Elasticsearch implementation of an `EventStore`, I found that I needed a composite key of `(timestamp, tiebreaker)` (where both `timestamp` and `tiebreaker` are `int64`s) to uniquely identify events, so a single `uint` isn't good enough (without relying on Postgres to generate auto-incrementing IDs).

I was experimenting with plugin-based `EventStore` implementations at the suggestion of @ari-becker, and realized that if the plugin runs in a separate process (i.e. we use something other than [Go plugins](https://golang.org/pkg/plugin/), which seem very limiting), we'll need some way of sending the `Key`s over the wire while retaining their ordering information. This isn't so easy with an arbitrary `Key` type, so I got back to thinking about restricting the `Key` type to something more concrete that has an implicit ordering.

From the Postgres and Elasticsearch implementations, I got the idea of using a `[]int64`. Using `timestamp` and `tiebreaker` could be used for many backends (the `timestamp` comes from the event, and `tiebreaker` is an in-memory counter - more details in the PR) - but for backends that support a more native keying approach (like Postgres' auto incrementing `event_id` or line numbers in file-based storage like S3 or local FS), we could make use of that instead. I wonder if there are backends that won't fit into the `[]int64` mold - if anyone has a system in mind, I'd love to hear about it. I also wonder if the `Key` could always be a `[2]int64`, or if there's value in having an arbitrary length slice.

Another approach is to avoid the `Key`/cursor approach altogether, which I'd definitely be open to ideas on.

Reply:
There's basically no way to establish sane global ordering without the whole nightmare of distributed consensus. I heartily discourage trying. Leave the business of sorting out ordering to consumers, who can choose what interpretations to apply.

But if what you need is _uniqueness_, that's much more straightforward. You have broadly two options:

1. UUIDs, specifically v4. Basically a 122-bit random number, so they can be generated without coordination. They are widely supported data type. PostgreSQL has a native type and the uuid-ossp module can generate them in the database very fast. They're supported in pretty much everything everywhere.
2. Natural keys. That is, you determine what the distinguishing fields are and then you digest them. This is isomorphic to the database normalisation exercise of identifying the natural key for a table: what is the field or combination of fields which absolutely defines uniqueness? In this case it might be something as simple as `(something_identifying_the_database, event_id)`. As long as the database is unique and the event_id is unique within the database, it "should" be unique overall.

I prefer option 1 to option 2. It's easier to implement and more robust to oversights. Natural keys are easy to describe but once you pick one, it's hard to fix if you discover you made a mistake and have collisions. Plus people are liable to start trying to rely on the natural key components, meaning you wouldn't be able to change it.

tl;dr UUIDs errywhere

Reply:
I think that's good advice - in my experimental PR for Elasticsearch, I made the naïve assumption that I could trust the timestamp on build events for ordering, when obviously isn't the case. Builds are tracked by a single ATC at a time, so at least time-based inconsistencies aren't as prevalent, but in general, it's a hacky "solution" (that doesn't really solve the problem). So, I agree that we shouldn't try to achieve distributed consensus.

> Leave the business of sorting out ordering to consumers, who can choose what interpretations to apply

This makes sense, but does this mean that Concourse will have to fetch all events for a build  (and sort them in memory) before it can serve any events? Or am I misunderstanding what you mean?

Another option that I suggested as a possibility is to use natural keys of `(build_id, event_id)`. For backends that don't support auto incrementing (e.g. Elasticsearch), rely on Postgres sequences to generate these `event_id`s. This way, we can take advantage of the indexing options of tools like Elasticsearch to perform sorting/pagination outside of Concourse.

Reply:
> This makes sense, but does this mean that Concourse will have to fetch all events for a build (and sort them in memory) before it can serve any events? Or am I misunderstanding what you mean?

Sorting in the database is fine, using the internal sequential IDs (ie the primary key). Those are typically monotonic, so they establish a local order. But they're only an ordinal scale (this-before-that), not an interval score (this-10-seconds-before-that) or ratio score (this-10-percent-longer-than-that).